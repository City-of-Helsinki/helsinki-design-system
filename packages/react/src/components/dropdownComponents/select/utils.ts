import React from 'react';

import { SelectData, SelectMetaData, SelectDataHandlers, FilterFunction } from './types';
import {
  Group,
  Option,
  ModularOptionListData,
  ModularOptionListProps,
  ScreenReaderNotification,
} from '../modularOptionList/types';
import { ChangeEvent } from '../../dataProvider/DataContext';
import { eventTypes } from './events';
import {
  iterateAndCopyGroup,
  getGroupLabelOption,
  getAllOptions,
  propsToGroups,
  getSelectedOptions,
} from '../modularOptionList/utils';

// Re-export commonly used utility functions
export {
  getAllOptions,
  getSelectedOptions,
  propsToGroups,
  updateGroupLabelAndOptions,
  updateOptionInGroup,
  defaultFilter,
} from '../modularOptionList/utils';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getElementIds(containerId: string): SelectMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: SelectData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function createScreenReaderNotification(type: string, content: string, delay = 0): ScreenReaderNotification {
  return {
    type,
    content,
    delay,
    showTime: 0,
    addTime: Date.now(),
  };
}

export function filterOptions(groups: ModularOptionListData['groups'], filterStr: string, filterFunc: FilterFunction) {
  const newGroupsWithFilteredOptions = iterateAndCopyGroup(groups, (option) => {
    if (option.isGroupLabel) {
      return { ...option };
    }
    return { ...option, visible: !filterStr || filterFunc(option, filterStr) };
  });

  // check if group label should be visible....
  newGroupsWithFilteredOptions.forEach((group) => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const groupLabel = getGroupLabelOption(group);
    if (!groupLabel) {
      return;
    }
    groupLabel.visible = !!groupLabel.label && group.options.findIndex((opt) => !opt.isGroupLabel && opt.visible) > -1;
  });

  return newGroupsWithFilteredOptions;
}

export function filterSelectableOptions(
  groups: ModularOptionListData['groups'],
  filterStr: string,
  filterFunc: FilterFunction,
  isMultiSelect: boolean,
) {
  return getAllOptions(groups, !isMultiSelect).filter((option) => filterFunc(option, filterStr));
}

export function mergeSearchResultsToCurrent(
  props: Pick<ModularOptionListProps, 'groups' | 'options'>,
  currentGroups: ModularOptionListData['groups'],
): ModularOptionListData['groups'] {
  const newData = propsToGroups(props) || [];
  const newOptions = getAllOptions(newData);
  const currentOptionsWithoutMatches = getSelectedOptions(currentGroups).filter((option) => {
    const sameInNewOptionsIndex = newOptions.findIndex((newOption) => {
      return newOption.value === option.value;
    });
    if (sameInNewOptionsIndex > -1) {
      newOptions[sameInNewOptionsIndex].selected = true;
      return false;
    }
    return true;
  });

  const currentHiddenOptionsInAGroup = currentOptionsWithoutMatches.length
    ? [{ options: currentOptionsWithoutMatches.map((opt) => ({ ...opt, visible: false })) } as Group]
    : [];

  return [...currentHiddenOptionsInAGroup, ...newData];
}

/**
 *
 * @param notification
 * @param dataHandlers
 * @returns Returns true, if notification was added
 */
export function addOrUpdateScreenReaderNotificationByType(
  notification: ScreenReaderNotification,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfSameType = screenReaderNotifications.findIndex((n) => n.type === notification.type);
  if (indexOfSameType > -1) {
    const updatedList = [...screenReaderNotifications];
    const prev = updatedList[indexOfSameType];
    if (prev.content === notification.content) {
      return false;
    }
    updatedList[indexOfSameType] = notification;
    dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
    return false;
  }
  const updatedList = [...screenReaderNotifications, notification];
  dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
  return true;
}

export function removeScreenReaderNotification(
  target: Partial<ScreenReaderNotification>,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfMatch = screenReaderNotifications.findIndex((n) => {
    const hasTypeMatch = !target.type || n.type === target.type;
    const hasContentMatch = !target.content || n.content === target.content;
    return hasTypeMatch && hasContentMatch;
  });
  if (indexOfMatch > -1) {
    screenReaderNotifications.splice(indexOfMatch, 1);
    dataHandlers.updateMetaData({ screenReaderNotifications });
    return true;
  }
  return false;
}
