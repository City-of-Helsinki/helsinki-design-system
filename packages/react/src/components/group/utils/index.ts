import {
  SyntheticEvent,
  FunctionComponent,
  HtmlHTMLAttributes,
  PropsWithChildren,
  ComponentType,
  MutableRefObject,
} from 'react';

import { StorageData, Storage } from './storage';

export type ChangeEventPayload = { value?: unknown; originalEvent?: SyntheticEvent };
export type ChangeEvent = { id: string; type?: string; payload?: ChangeEventPayload };
export type ChangeHandlerProps = ChangeEvent & { controller: Controller };
export type ChangeHandler = (props: ChangeHandlerProps) => unknown;
export type HTMLElementAttributesWithChildren = PropsWithChildren<HtmlHTMLAttributes<HTMLElement>>;
export type DefaultGroupElementProps = Record<string, unknown> & {
  key?: string;
  forwardedController?: Controller;
  'data-hds-group-id'?: string;
};

export type ElementRef = MutableRefObject<HTMLElement | null>;

export type Controller = {
  getProps: ({ id: string, elementProps: DefaultGroupElementProps }) => ReturnType<PropSetter>;
  updateData: (props: { data: StorageData; updateKeys?: string[] }) => void;
  getData: Storage['get'];
  getMetaData: () => StorageData;
  updateMetaData: (appended: StorageData) => StorageData;
  triggerChange: (change: ChangeEvent) => unknown;
};

export type ControllerProps = {
  reRenderer: () => void;
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange: ChangeHandler;
};

export type PropSetter<T = DefaultGroupElementProps> = (props: {
  id: string;
  controller: Controller;
  key: string;
  elementProps: DefaultGroupElementProps;
}) => T;

export type PropHandler = {
  getProps: (props: {
    id: string;
    controller: Controller;
    elementProps: DefaultGroupElementProps;
  }) => DefaultGroupElementProps;
  updateKey: (id: string) => void;
};

export type GroupChildFunction = FunctionComponent<{ controller: Controller }>;

export type GroupChild<P = DefaultGroupElementProps> = ComponentType<HTMLElementAttributesWithChildren & P>;

export type GroupProps = PropsWithChildren<unknown> & {
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange?: ChangeHandler;
};
