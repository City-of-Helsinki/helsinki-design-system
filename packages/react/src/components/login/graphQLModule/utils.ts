import { QueryOptions } from '@apollo/client';
import { merge } from 'lodash';

import { GraphQLModuleModuleProps } from '.';
import { ApiTokenClient } from '../apiTokensClient';

type PartialContext = Partial<QueryOptions['context']>;
type QueryHeaders = Record<string, unknown>;

function mergeQueryOptionsContexts(target: PartialContext, overrides: PartialContext): PartialContext {
  return merge(target, overrides);
}

export function mergeQueryOptionContexts(options: QueryOptions, overrideContext: PartialContext): QueryOptions {
  const context: PartialContext = options.context || {};
  // eslint-disable-next-line no-param-reassign
  options.context = mergeQueryOptionsContexts(context, overrideContext);
  return options;
}

export function mergeHeadersToQueryOptions(options: QueryOptions, headers: QueryHeaders): QueryOptions {
  const context: PartialContext = {
    headers,
  };
  return mergeQueryOptionContexts(options, context);
}

export function mergeAuthorizationHeaderToQueryOptions(options: QueryOptions, authHeader: string): QueryOptions {
  const headers: QueryHeaders = {
    authorization: authHeader,
  };
  return mergeHeadersToQueryOptions(options, headers);
}

export function mergeQueryOptions(
  target: Partial<GraphQLModuleModuleProps['queryOptions']>,
  overrides: Partial<GraphQLModuleModuleProps['queryOptions']>,
) {
  return merge(target, overrides);
}

/* No need to export to hds-react package */
export function mergeQueryOptionsToModuleProps<T, Q>(
  target: Partial<GraphQLModuleModuleProps<T, Q>>,
  queryOptions: Partial<QueryOptions<Q>>,
) {
  const { query, ...rest } = queryOptions;
  // eslint-disable-next-line no-param-reassign
  target.queryOptions = mergeQueryOptions(target.queryOptions || {}, rest);
  if (query) {
    // eslint-disable-next-line no-param-reassign
    target.query = query;
  }
  return target;
}

export function setBearerToQueryOptions(queryOptions: QueryOptions, token: string) {
  if (!token) {
    return queryOptions;
  }
  return mergeAuthorizationHeaderToQueryOptions(queryOptions, `Bearer ${token}`);
}

export function appendFetchOptions(queryOptions: QueryOptions, newOptions: Partial<Parameters<typeof fetch>[1]>) {
  if (!queryOptions.context) {
    // eslint-disable-next-line no-param-reassign
    queryOptions.context = {};
  }

  // eslint-disable-next-line no-param-reassign
  queryOptions.context.fetchOptions = {
    ...queryOptions.context.fetchOptions,
    ...newOptions,
  };
  return queryOptions;
}

/* No need to export to hds-react package */
export function mergeQueryOptionModifiers({
  options,
  queryHelper,
}: Pick<GraphQLModuleModuleProps, 'queryHelper' | 'options'>): Required<GraphQLModuleModuleProps>['queryHelper'] {
  const { apiTokenKey } = options || {};
  if (!apiTokenKey && !queryHelper) {
    return (opt: QueryOptions) => opt;
  }

  if (!apiTokenKey && queryHelper) {
    return queryHelper;
  }

  const tokenPicker = (apiTokenClient: ApiTokenClient, key: string) => {
    if (!apiTokenClient) {
      return undefined;
    }
    const tokens = apiTokenClient.getTokens();
    return tokens ? tokens[key] : undefined;
  };

  const apiTokenSetter: GraphQLModuleModuleProps['queryHelper'] = (opt, apiTokenClient) => {
    if (!apiTokenKey || !apiTokenClient) {
      return opt;
    }
    const token = tokenPicker(apiTokenClient, apiTokenKey);
    if (token) {
      return setBearerToQueryOptions(opt, token);
    }
    return opt;
  };

  if (!queryHelper) {
    return apiTokenSetter;
  }

  return (opt, apiTokenClient, beacon) => {
    return queryHelper(apiTokenSetter(opt, apiTokenClient, beacon), apiTokenClient, beacon);
  };
}
