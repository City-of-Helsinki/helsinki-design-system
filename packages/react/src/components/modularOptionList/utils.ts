import React, { ReactElement, ReactNode } from 'react';

import {
  ModularOptionListData,
  Group,
  ModularOptionListProps,
  Option,
  OptionInProps,
  ModularOptionListMetaData,
  GroupInProps,
  ScreenReaderNotification,
  ModularOptionListDataHandlers,
} from './types';
import { getChildrenAsArray } from '../../utils/getChildren';
import { ChangeEvent } from '../dataProvider/DataContext';
import { eventTypes } from './events';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getOptionIndex(options: OptionInProps[], option: OptionInProps): number {
  return options.findIndex(
    ({ value, isGroupLabel }) => value === option.value && !!isGroupLabel === !!option.isGroupLabel,
  );
}

export function getOptionGroupIndex(groups: ModularOptionListData['groups'], option: OptionInProps): number {
  if (groups.length === 0) {
    return -1;
  }
  if (groups.length === 1) {
    return 0;
  }
  return groups.findIndex(({ options }) => {
    return getOptionIndex(options, option) > -1;
  });
}

export function iterateAndCopyGroup(groups: Group[], iterator: OptionIterator): Group[] {
  return groups.map((group, groupIndex) => {
    return {
      options: group.options.map((opt, optionIndex) => {
        return iterator(opt, group, optionIndex, groupIndex) || { ...opt };
      }),
    };
  });
}

export function getSelectedOptionsPerc(group: Group, pendingSelectionCount = 0): number {
  const optionCountWithoutGroupLabel = group.options.length - 1;
  if (!optionCountWithoutGroupLabel) {
    return 0;
  }
  return (
    (group.options.filter((option) => !option.isGroupLabel && option.selected).length + pendingSelectionCount) /
    optionCountWithoutGroupLabel
  );
}

export function mutateGroupLabelSelections(groups: ModularOptionListData['groups']) {
  groups.forEach((group) => {
    // eslint-disable-next-line no-param-reassign
    group.options[0].selected = getSelectedOptionsPerc(group) === 1;
  });

  return groups;
}

// this function mutates given data so pass a copy of groups and their options array.
// options are not mutated.
function mutateGroupsWithOptions(
  groups: ModularOptionListData['groups'],
  partialOptionsToUpdate: Partial<Option>[],
  isMultiSelect: boolean,
) {
  partialOptionsToUpdate.forEach((partial) => {
    if (!partial.value) {
      return;
    }
    const groupIndex = getOptionGroupIndex(groups, { ...partial, isGroupLabel: !!partial.isGroupLabel });
    if (groupIndex < 0) {
      return;
    }
    const group = groups[groupIndex];
    const optionIndex = getOptionIndex(group.options, partial);
    const option = optionIndex > -1 ? group.options[optionIndex] : undefined;
    if (option) {
      group.options[optionIndex] = {
        ...option,
        ...partial,
      };
    }
  });

  if (isMultiSelect) {
    mutateGroupLabelSelections(groups);
  }
  return groups;
}

export function updateOptionInGroup(
  groups: ModularOptionListData['groups'],
  updatedOption: Option,
  isMultiSelect: boolean,
): ModularOptionListData['groups'] {
  if (updatedOption.isGroupLabel) {
    throw new Error('Use updateGroupLabelAndOptions to update groupLabel and its related options');
  }

  const copy = iterateAndCopyGroup(groups, (opt) => {
    // unselect others if single select and option is selected
    if (!isMultiSelect && !opt.isGroupLabel && updatedOption.selected) {
      return {
        ...opt,
        selected: false,
      };
    }
    return { ...opt };
  });

  return mutateGroupsWithOptions(copy, [{ ...updatedOption, selected: !!updatedOption.selected }], isMultiSelect);
}

export function updateGroupLabelAndOptions(
  groups: ModularOptionListData['groups'],
  updatedOption: Option,
): ModularOptionListData['groups'] {
  if (!updatedOption.isGroupLabel) {
    throw new Error('Use updateOptionInGroup to update non-groupLabel options.');
  }
  if (!updatedOption.visible) {
    throw new Error('Cannot click an group label that is not visible (without a label)');
  }
  const targetGroupIndex = getOptionGroupIndex(groups, updatedOption);
  if (targetGroupIndex < 0) {
    return groups;
  }

  return iterateAndCopyGroup(groups, (option, group, optionIndex, groupIndex) => {
    if (groupIndex !== targetGroupIndex) {
      return { ...option };
    }
    return option.visible && !option.disabled
      ? {
          ...option,
          selected: updatedOption.selected,
        }
      : { ...option };
  });
}

export function clearAllSelectedOptions(
  groups: ModularOptionListData['groups'],
  keepDisabled = true,
): ModularOptionListData['groups'] {
  return iterateAndCopyGroup(groups, (option) => {
    return {
      ...option,
      selected: option.disabled && keepDisabled ? option.selected : false,
    };
  });
}

export function getAllOptions(groups: ModularOptionListData['groups'], filterOutGroupLabels = true): Option[] {
  const options: Option[] = [];
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (filterOutGroupLabels && option.isGroupLabel) {
        return;
      }
      options.push(option);
    });
  });
  return options;
}

export function getVisibleGroupLabels(groups: ModularOptionListData['groups']): Option[] {
  return groups.map((group) => group.options[0]).filter((option) => option && option.label && option.visible);
}

export function getSelectedOptions(groups: ModularOptionListData['groups']): Option[] {
  return getAllOptions(groups).filter((option) => !!option.selected);
}

export function validateOption(option: OptionInProps | string): Option {
  if (typeof option === 'string') {
    return { value: option, label: option, selected: false, isGroupLabel: false, visible: true, disabled: false };
  }

  const label = option.label || option.value || '';
  const value = option.value || label;

  return {
    label,
    value,
    selected: !!option.selected,
    isGroupLabel: false,
    visible: typeof option.visible === 'boolean' ? option.visible : true,
    disabled: typeof option.disabled === 'boolean' ? option.disabled : false,
  };
}

export function createGroupLabel(label: string) {
  return { ...validateOption(String(label)), isGroupLabel: true, visible: !!label, disabled: false };
}

export function propsToGroups(
  props: Pick<ModularOptionListProps, 'groups' | 'options'>,
): ModularOptionListData['groups'] | undefined {
  if (!props.groups && !props.options) {
    return undefined;
  }
  if (props.groups) {
    return (props.groups as Group[]).map((group: Group) => {
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      const hasLabelOptionAlready = !!getGroupLabelOption(group);
      const groupOptions = group.options.map(validateOption) as Option[];
      if (hasLabelOptionAlready) {
        return {
          options: [group.options[0], ...groupOptions.slice(1)],
        };
      }
      const labelOption: Option = createGroupLabel((group as GroupInProps).label);

      return {
        options: [labelOption, ...groupOptions],
      };
    });
  }

  if (props.options) {
    const option = props.options[0];
    if (option && (option as OptionInProps).isGroupLabel) {
      return [props] as Group[];
    }
  }

  return [
    {
      options: [createGroupLabel(''), ...(props.options || []).map(validateOption)],
    },
  ];
}

export function childrenToGroups(
  children: ModularOptionListProps['children'],
): ModularOptionListData['groups'] | undefined {
  if (!children || typeof children !== 'object') {
    return undefined;
  }
  const childArray = getChildrenAsArray(children) as ReactElement[];
  if (!childArray.length) {
    return [{ options: [] }];
  }
  const hasOptionGroups = childArray[0].type === 'optgroup';
  const optionElementToOption = (optionEl: ReactNode | ReactElement) => {
    const props = (
      optionEl && typeof optionEl === 'object' ? (optionEl as ReactElement).props : {}
    ) as ReactElement<HTMLOptionElement>['props'];
    const label = String(props.children);
    const value = props && String(props.value);
    const selected = !!(props && props.selected);
    const disabled = !!(props && props.disabled);
    return validateOption({ label, value, selected, disabled });
  };
  if (hasOptionGroups) {
    return childArray.map((child) => {
      const optionElements = child.props.children;
      const options = optionElements ? getChildrenAsArray(optionElements).map(optionElementToOption) : [];
      const label = createGroupLabel(String(child.props.label));
      options.unshift(label);
      return { options };
    });
  }
  return [{ options: [createGroupLabel(''), ...childArray.map(optionElementToOption)] }];
}

export function createSelectedOptionsList(currentSelections: Option[], groups: ModularOptionListData['groups']) {
  const selections = getSelectedOptions(groups);
  const selectionsAsValues = new Set(selections.map((opt) => opt.value));
  const stillSelected = currentSelections.filter((opt) => selectionsAsValues.has(opt.value));
  const selectedValues = new Set(stillSelected.map((opt) => opt.value));
  return [...stillSelected, ...selections.filter((opt) => !selectedValues.has(opt.value))];
}

export function getElementIds(containerId: string): ModularOptionListMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: ModularOptionListData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function getGroupLabelOption(group: Group): Option | undefined {
  const firstOption = group.options[0];
  return firstOption && firstOption.isGroupLabel ? firstOption : undefined;
}

export function createScreenReaderNotification(type: string, content: string, delay = 0): ScreenReaderNotification {
  return {
    type,
    content,
    delay,
    showTime: 0,
    addTime: Date.now(),
  };
}

/**
 *
 * @param notification
 * @param dataHandlers
 * @returns Returns true, if notification was added
 */
export function addOrUpdateScreenReaderNotificationByType(
  notification: ScreenReaderNotification,
  dataHandlers: ModularOptionListDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfSameType = screenReaderNotifications.findIndex((n) => n.type === notification.type);
  if (indexOfSameType > -1) {
    const updatedList = [...screenReaderNotifications];
    const prev = updatedList[indexOfSameType];
    if (prev.content === notification.content) {
      return false;
    }
    updatedList[indexOfSameType] = notification;
    dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
    return false;
  }
  const updatedList = [...screenReaderNotifications, notification];
  dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
  return true;
}

export function removeScreenReaderNotification(
  target: Partial<ScreenReaderNotification>,
  dataHandlers: ModularOptionListDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfMatch = screenReaderNotifications.findIndex((n) => {
    const hasTypeMatch = !target.type || n.type === target.type;
    const hasContentMatch = !target.content || n.content === target.content;
    return hasTypeMatch && hasContentMatch;
  });
  if (indexOfMatch > -1) {
    screenReaderNotifications.splice(indexOfMatch, 1);
    dataHandlers.updateMetaData({ screenReaderNotifications });
    return true;
  }
  return false;
}

export function findSelectableOptionIndex(
  groups: ModularOptionListData['groups'],
  iterator: (option: Option) => boolean,
  isMultiSelect: boolean,
) {
  return getAllOptions(groups, !isMultiSelect).findIndex(iterator);
}

export function pickSelectedValues(selectedOptions?: Option[]): string[] {
  if (!selectedOptions || !selectedOptions.length) {
    return [];
  }
  return selectedOptions.map((opt) => opt.value);
}

export function getNewSelections(prev: Option[], current: Option[]): Option[] {
  if (!prev.length) {
    return current;
  }
  const prevValues = pickSelectedValues(prev);
  return current.filter((opt) => !prevValues.includes(opt.value));
}

// this function does not update groupLabel's selected with multiselect,
// because isMultiSelect does not exist where this is used.
// use mutateGroupLabelSelections to update group label selections too.
export function updateSelectedOptionsInGroups(
  groups: ModularOptionListProps['groups'],
  selectedOptions?: Array<Option | OptionInProps | string> | string,
) {
  if (selectedOptions === undefined) {
    return groups;
  }
  const selectedOptionsAsOptions = (typeof selectedOptions === 'string' ? [selectedOptions] : selectedOptions).map(
    (opt) => {
      const { value } = validateOption(opt);
      return { value, selected: true };
    },
  );
  // passing groups as groups does not matter
  const copy = iterateAndCopyGroup(groups as Group[], (opt) => {
    // unselect others if single select and option is selected
    const validOption = typeof opt === 'string' ? validateOption(opt) : opt;

    return {
      ...validOption,
      selected: false,
    };
  });
  // keep labels if groups were GroupInProps
  (groups as GroupInProps[]).forEach((group, index) => {
    if (group.label) {
      (copy[index] as GroupInProps).label = group.label;
    }
  });
  return mutateGroupsWithOptions(copy, selectedOptionsAsOptions, false);
}

export function convertPropsToGroups({
  groups,
  options,
  value,
  children,
}: Pick<ModularOptionListProps, 'groups' | 'options' | 'value' | 'children'>): Group[] {
  const fromGroupsAndOptions = propsToGroups({ options, groups });
  if (fromGroupsAndOptions) {
    if (value) {
      const selectedOptions = getSelectedOptions(fromGroupsAndOptions);
      if (selectedOptions.length > 0) {
        // eslint-disable-next-line no-console
        console.warn('HDS Select component has both selected options and value set. Value is discarded');
        return fromGroupsAndOptions;
      }
      return updateSelectedOptionsInGroups(fromGroupsAndOptions, value) as Group[];
    }
    return fromGroupsAndOptions;
  }
  return childrenToGroups(children) || [];
}

export function createMetaDataAfterSelectionChange(
  groups: ModularOptionListData['groups'],
  selectedOptions: ModularOptionListMetaData['selectedOptions'],
  lastClickedOption?: Option,
): Partial<Pick<ModularOptionListMetaData, 'selectedOptions' | 'textContent' | 'lastClickedOption'>> {
  const data: Partial<ModularOptionListMetaData> = {
    selectedOptions: createSelectedOptionsList(selectedOptions, groups),
    lastClickedOption,
    // textContent is re-created, when a textProvider is called
    textContent: undefined,
  };

  return data;
}
