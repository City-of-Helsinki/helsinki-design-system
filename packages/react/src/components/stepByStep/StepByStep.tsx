import React, { FC } from 'react';

import '../../styles/base.module.css';
import styles from './StepByStep.module.scss';
import { Button, ButtonProps } from '../button';
import classNames from '../../utils/classNames';
import { Link, LinkProps } from '../link';
import { AllElementPropsWithoutRef } from '../../utils/elementTypings';

type StepType = {
  /**
   * List of native button objects for the step.
   */
  buttons?: Array<ButtonProps>;
  /**
   * Step description.
   */
  description?: JSX.Element | string;
  /**
   * Optional key for the React component. Title is used unless a key is provided.
   */
  // eslint-disable-next-line react/no-unused-prop-types
  key?: string;
  /**
   * List of native link objects for the step.
   */
  links?: Array<LinkProps>;
  /**
   * Step title.
   */
  title: string;
};

export type StepByStepProps = AllElementPropsWithoutRef<'div'> & {
  /**
   * Additional class names to apply to the container element.
   */
  className?: string;

  /**
   * Class name for the main heading.
   */
  headerClassName?: string;

  /**
   * Header level for the main heading.
   * @default 2
   */
  headerLevel?: number;

  /**
   * Help text for the step by step component. Displayed under the `title` property.
   */
  helpText?: string;

  /**
   * Boolean indicating whether the steps are numbered.
   */
  numberedList?: boolean;

  /**
   * Steps for the step by step component.
   */
  steps: Array<StepType>;

  /**
   * Title text for the step by step component.
   */
  title?: string;
};

const resolveElementKey = (props: ButtonProps | LinkProps) => {
  if (props.key) {
    return props.key;
  }
  if (typeof props.children === 'string') {
    return props.children as string;
  }
  return undefined;
};

const getButtonOrLinkRenderer = (Tag: typeof Button | typeof Link) => (componentProps: ButtonProps & LinkProps) => {
  const { children, ...props } = componentProps;
  return (
    <p key={resolveElementKey(componentProps)}>
      <Tag {...props}>{children}</Tag>
    </p>
  );
};

const renderLink = getButtonOrLinkRenderer(Link);
const renderButton = getButtonOrLinkRenderer(Button);

const StepComponent = ({ title, description, buttons = [], links = [] }: StepType) => {
  return (
    <li className={styles.stepItem}>
      <p className={styles.stepItemTitle}>{title}</p>
      <div>
        {description && <p key="description">{description}</p>}
        {buttons.map(renderButton)}
        {links.map(renderLink)}
      </div>
    </li>
  );
};

export const StepByStep: FC<StepByStepProps> = ({
  className,
  title,
  helpText,
  steps = [],
  numberedList = false,
  headerClassName,
  headerLevel = 2,
  ...rest
}) => {
  const wrapperClassName = classNames(styles.container, className);
  const titleComponent =
    title && React.createElement(`h${headerLevel}`, { className: classNames(styles.title, headerClassName) }, title);

  return (
    <div {...rest} className={wrapperClassName}>
      <div>
        {titleComponent}
        <p className={styles.description}>{helpText}</p>
      </div>
      {React.createElement(
        numberedList ? 'ol' : 'ul',
        { className: styles.stepsContainer },
        steps.map((step) => <StepComponent key={step.key || step.title} {...step} />),
      )}
    </div>
  );
};
