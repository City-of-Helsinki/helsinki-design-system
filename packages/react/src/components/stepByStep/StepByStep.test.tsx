import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { StepByStep } from './StepByStep';
import { getCommonElementTestProps, getElementAttributesMisMatches } from '../../utils/testHelpers';

const steps = [
  {
    title: 'Vaiheen 1 otsikko',
    description:
      'Tähän voit lisätä tekstiä, joka kertoo käyttäjälle mitä kyseisessä vaiheessa tapahtuu. Pidä teksti tiiviinä, jotta käyttäjä saa kokonaiskuvan prosessista ja sen vaiheista helposti silmäilemällä.',
    buttons: [
      {
        children: 'Esimerkki painikkeesta',
        href: 'https://hel.fi',
      },
    ],
    links: [
      {
        children: 'Esimerkki lisätietolinkistä',
        href: 'https://hel.fi',
      },
    ],
  },
  {
    title: 'Vaiheen 2 otsikko',
    description: 'Tähän voit lisätä tekstiä.',
  },
];

describe('<StepByStep /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<StepByStep numberedList steps={steps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(<StepByStep steps={steps} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  it('native html props are passed to the element', async () => {
    const divProps = getCommonElementTestProps('div');
    const { getByTestId } = render(<StepByStep {...divProps} steps={steps} />);
    const element = getByTestId(divProps['data-testid']);
    expect(getElementAttributesMisMatches(element, { ...divProps })).toHaveLength(0);
  });
});
