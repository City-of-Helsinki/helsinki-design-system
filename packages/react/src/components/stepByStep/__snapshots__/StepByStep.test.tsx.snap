// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepByStep /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div>
      <p
        class="description"
      />
    </div>
    <ol
      class="stepsContainer"
    >
      <li
        class="stepItem"
      >
        <p
          class="stepItemTitle"
        >
          Vaiheen 1 otsikko
        </p>
        <div>
          <p>
            Tähän voit lisätä tekstiä, joka kertoo käyttäjälle mitä kyseisessä vaiheessa tapahtuu. Pidä teksti tiiviinä, jotta käyttäjä saa kokonaiskuvan prosessista ja sen vaiheista helposti silmäilemällä.
          </p>
          <p>
            <button
              class="button primary theme-bus size-medium"
              href="https://hel.fi"
              type="button"
            >
              <span>
                Esimerkki painikkeesta
              </span>
            </button>
          </p>
          <p>
            <a
              class="link link-medium"
              href="https://hel.fi"
            >
              Esimerkki lisätietolinkistä
            </a>
          </p>
        </div>
      </li>
      <li
        class="stepItem"
      >
        <p
          class="stepItemTitle"
        >
          Vaiheen 2 otsikko
        </p>
        <div>
          <p>
            Tähän voit lisätä tekstiä.
          </p>
        </div>
      </li>
    </ol>
  </div>
</DocumentFragment>
`;
